include/contraction/contractionGraph.hpp

```diff
@@ -40,220 +40,304 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+            for (boost::tie(out, out_end) = out_edges(*vi, g.graph);
+                    out != out_end; ++out) {
+                os << ' ' << g.graph[*out].id
+                    << "=(" << g.graph[g.source(*out)].id
+                    << ", " << g.graph[g.target(*out)].id << ") = "
+                    <<  g.graph[*out].cost <<"\t";
+            }
+            os << std::endl;
+        }
+        return os;
+    }
+
+    /*!
+        @brief add a shortcut to the graph during contraction
+        @param [in] edge of type *CH_edge* is to be added
+        @param [in] u vertex
+        @param [in] v vertex
+        u -> w -> v
+        u -> v
+        edge (u, v) is a new edge e
+        contracted_vertices = w + contracted vertices
+    */
+    void add_shortcut(const CH_edge &edge, V u, V v) {
+        bool inserted;
+        E e;
+        if (edge.cost < 0) return;
+        boost::tie(e, inserted) = boost::add_edge(u, v, this->graph);
+        this->graph[e]= edge;
+    }
+
+    /*
+     *
+     * u ----e1{v1}----> v ----e2{v2}----> w
+     *
+     * e1: min cost edge from u to v
+     * e2: min cost edge from v to w
+     *
+     * result:
+     * u ---{v+v1+v2}---> w
+     *
+     */
+    void process_shortcut(V u, V v, V w) {
+        auto e1 = get_min_cost_edge(u, v);
+        auto e2 = get_min_cost_edge(v, w);
+
+        if (std::get<2>(e1) && std::get<2>(e2)) {
+            auto contracted_vertices = std::get<1>(e1) + std::get<1>(e2);
+            double cost = std::get<0>(e1) + std::get<0>(e2);
+            contracted_vertices += (this->graph[v]).id;
+            contracted_vertices += (this->graph[v]).get_contracted_vertices();
+
+            // Create shortcut
+            CH_edge shortcut(
+                get_next_id(),
+                (this->graph[u]).id,
+                (this->graph[w]).id,
+                cost);
+            shortcut.set_contracted_vertices(contracted_vertices);
+            add_shortcut(shortcut, u, w);
+        }
+    }
+
+    /*! 
+        @brief tests if the edges sequence (u, v), (v, w) exists in the graph
+    */
+    bool has_u_v_w(V u, V v, V w) const {
+        return boost::edge(u, v, this->graph).second
+            && boost::edge(v, w, this->graph).second;
+    }
+
+    /*
+        Possibility of a shortcut from left vertex to right vertex
+        *v* should be a linear vertex
+        u <-> v -> w: v not considered linear
+        @dot
+        graph G {
+        graph [rankdir=LR];
+        subgraph cluster0 {
+        node [shape=point,height=0.2,style=filled,color=black];
+        style=filled;
+        color=lightgrey;
+        a0; a1; a2;
+        label = "rest of graph";
+        }
+        v [color=green];
+        v -- left;
+        v -- right;
+        u -- a0;
+        w -- a1;
+        }
+        @enddot
+    */
+    bool is_shortcut_possible(V u, V v, V w) {
+        if (u == v || v == w || u == w) return false;
+        pgassert(u != v);
```

Comment:
not possible?

Reply:
No, we cannot build shortcuts between u and w avoiding u.
u, v and w must be distinct.